# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    eos = source_vocab_to_int['<EOS>']
    source = [[source_vocab_to_int[word] for word in sentence.split()] for sentence in source_text.split('\n')]
    target = [[target_vocab_to_int[word] for word in sentence.split()] + [eos] for sentence in target_text.split('\n')]
    return source, target

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.3.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [8]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='target')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None], name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    return inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main\n    "__main__", mod_spec)', 'File "/usr/lib/python3.5/runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "/home/gao/.local/lib/python3.5/site-packages/ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "/home/gao/.local/lib/python3.5/site-packages/traitlets/config/application.py", line 658, in launch_instance\n    app.start()', 'File "/home/gao/.local/lib/python3.5/site-packages/ipykernel/kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "/home/gao/.local/lib/python3.5/site-packages/zmq/eventloop/ioloop.py", line 177, in start\n    super(ZMQIOLoop, self).

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [9]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    go = source_vocab_to_int['<GO>']
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    decoded = tf.concat([tf.fill([batch_size, 1], go), ending], 1)
    return decoded

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [13]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    
    encoder_embedding = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    encoder_cell = tf.contrib.rnn.MultiRNNCell([ tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.LSTMCell(rnn_size,
                                                                        initializer = tf.random_uniform_initializer(-0.1, 0.1, 
                                                                        seed = 2017)), keep_prob) for x in range(num_layers)])
        
    rnn_output, rnn_state = tf.nn.dynamic_rnn(encoder_cell, 
                                              encoder_embedding, 
                                              sequence_length = source_sequence_length,
                                              dtype = tf.float32)
    return rnn_output, rnn_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [14]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    train_helper = tf.contrib.seq2seq.TrainingHelper(inputs = dec_embed_input,
                                                              sequence_length = target_sequence_length,
                                                              time_major = False)
    train_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, train_helper, encoder_state, output_layer)
    return tf.contrib.seq2seq.dynamic_decode(train_decoder, 
                                             impute_finished = True, 
                                             maximum_iterations= max_summary_length)[0]



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [15]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype = tf.int32), [batch_size], name = 'start_tokens')
    
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                               start_tokens,
                                                               end_of_sequence_id)
    
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                       inference_helper,
                                                       encoder_state,
                                                       output_layer)
    
    return tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                             impute_finished = True,
                                             maximum_iterations = max_target_sequence_length)[0]



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [16]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    decoder_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    decoder_embed_input = tf.nn.embedding_lookup(decoder_embeddings, dec_input)

    decoder_cell = tf.contrib.rnn.MultiRNNCell([ tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.LSTMCell(rnn_size,
                                            initializer=tf.random_uniform_initializer(-0.1,0.1,seed=2007)), keep_prob) for _ in range(num_layers)])
    
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    with tf.variable_scope("decode"):
        train_decoder_out = decoding_layer_train(encoder_state, 
                                                 decoder_cell, 
                                                 decoder_embed_input, 
                                                 target_sequence_length, 
                                                 max_target_sequence_length, 
                                                 output_layer, 
                                                 keep_prob)
        
    with tf.variable_scope("decode", reuse=True):
        infer_decoder_out = decoding_layer_infer(encoder_state, 
                                                 decoder_cell, 
                                                 decoder_embeddings, 
                                                 target_vocab_to_int['<GO>'], 
                                                 target_vocab_to_int['<EOS>'], 
                                                 max_target_sequence_length, 
                                                 target_vocab_size, 
                                                 output_layer, 
                                                 batch_size, keep_prob)
        
    return (train_decoder_out, infer_decoder_out)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [17]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    _, enc_state = encoding_layer(input_data, 
                               rnn_size, 
                               num_layers, 
                               keep_prob, 
                               source_sequence_length, 
                               source_vocab_size, 
                               enc_embedding_size)
    
    proc_data = process_decoder_input(target_data, target_vocab_to_int, batch_size) 
    return decoding_layer(proc_data, 
                          enc_state,
                          target_sequence_length,
                          max_target_sentence_length,
                          rnn_size,
                          num_layers,
                          target_vocab_to_int,
                          target_vocab_size,
                          batch_size,
                          keep_prob,
                          dec_embedding_size)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [18]:
# Number of Epochs
epochs = 16
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 5

### Build the Graph
Build the graph using the neural network you implemented.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    5/538 - Train Accuracy: 0.3015, Validation Accuracy: 0.3542, Loss: 3.4206
Epoch   0 Batch   10/538 - Train Accuracy: 0.3047, Validation Accuracy: 0.3885, Loss: 3.1383
Epoch   0 Batch   15/538 - Train Accuracy: 0.4258, Validation Accuracy: 0.4494, Loss: 2.7369
Epoch   0 Batch   20/538 - Train Accuracy: 0.4128, Validation Accuracy: 0.4553, Loss: 2.5938
Epoch   0 Batch   25/538 - Train Accuracy: 0.4197, Validation Accuracy: 0.4714, Loss: 2.5114
Epoch   0 Batch   30/538 - Train Accuracy: 0.4213, Validation Accuracy: 0.4837, Loss: 2.4135
Epoch   0 Batch   35/538 - Train Accuracy: 0.4186, Validation Accuracy: 0.4764, Loss: 2.2631
Epoch   0 Batch   40/538 - Train Accuracy: 0.5044, Validation Accuracy: 0.5036, Loss: 1.9836
Epoch   0 Batch   45/538 - Train Accuracy: 0.4903, Validation Accuracy: 0.5057, Loss: 1.9606
Epoch   0 Batch   50/538 - Train Accuracy: 0.4703, Validation Accuracy: 0.5057, Loss: 1.8985
Epoch   0 Batch   55/538 - Train Accuracy: 0.4584, Validation Accuracy

Epoch   0 Batch  450/538 - Train Accuracy: 0.7883, Validation Accuracy: 0.7591, Loss: 0.3439
Epoch   0 Batch  455/538 - Train Accuracy: 0.8137, Validation Accuracy: 0.7786, Loss: 0.3002
Epoch   0 Batch  460/538 - Train Accuracy: 0.7755, Validation Accuracy: 0.7766, Loss: 0.3086
Epoch   0 Batch  465/538 - Train Accuracy: 0.7937, Validation Accuracy: 0.7594, Loss: 0.3175
Epoch   0 Batch  470/538 - Train Accuracy: 0.7954, Validation Accuracy: 0.7670, Loss: 0.2984
Epoch   0 Batch  475/538 - Train Accuracy: 0.7999, Validation Accuracy: 0.7775, Loss: 0.3002
Epoch   0 Batch  480/538 - Train Accuracy: 0.7818, Validation Accuracy: 0.7662, Loss: 0.3039
Epoch   0 Batch  485/538 - Train Accuracy: 0.8032, Validation Accuracy: 0.7770, Loss: 0.2895
Epoch   0 Batch  490/538 - Train Accuracy: 0.8065, Validation Accuracy: 0.7839, Loss: 0.2853
Epoch   0 Batch  495/538 - Train Accuracy: 0.8029, Validation Accuracy: 0.7976, Loss: 0.2898
Epoch   0 Batch  500/538 - Train Accuracy: 0.8143, Validation Accuracy

Epoch   1 Batch  360/538 - Train Accuracy: 0.9086, Validation Accuracy: 0.8988, Loss: 0.1032
Epoch   1 Batch  365/538 - Train Accuracy: 0.8590, Validation Accuracy: 0.8908, Loss: 0.0981
Epoch   1 Batch  370/538 - Train Accuracy: 0.9119, Validation Accuracy: 0.8890, Loss: 0.1028
Epoch   1 Batch  375/538 - Train Accuracy: 0.9083, Validation Accuracy: 0.8904, Loss: 0.1000
Epoch   1 Batch  380/538 - Train Accuracy: 0.9187, Validation Accuracy: 0.9032, Loss: 0.0945
Epoch   1 Batch  385/538 - Train Accuracy: 0.9077, Validation Accuracy: 0.8908, Loss: 0.0981
Epoch   1 Batch  390/538 - Train Accuracy: 0.9269, Validation Accuracy: 0.8924, Loss: 0.0895
Epoch   1 Batch  395/538 - Train Accuracy: 0.8920, Validation Accuracy: 0.8912, Loss: 0.1047
Epoch   1 Batch  400/538 - Train Accuracy: 0.9033, Validation Accuracy: 0.8888, Loss: 0.0966
Epoch   1 Batch  405/538 - Train Accuracy: 0.8824, Validation Accuracy: 0.8880, Loss: 0.0910
Epoch   1 Batch  410/538 - Train Accuracy: 0.8980, Validation Accuracy

Epoch   2 Batch  270/538 - Train Accuracy: 0.9211, Validation Accuracy: 0.9039, Loss: 0.0619
Epoch   2 Batch  275/538 - Train Accuracy: 0.9205, Validation Accuracy: 0.9164, Loss: 0.0822
Epoch   2 Batch  280/538 - Train Accuracy: 0.9293, Validation Accuracy: 0.9084, Loss: 0.0604
Epoch   2 Batch  285/538 - Train Accuracy: 0.9321, Validation Accuracy: 0.9059, Loss: 0.0567
Epoch   2 Batch  290/538 - Train Accuracy: 0.9309, Validation Accuracy: 0.9009, Loss: 0.0657
Epoch   2 Batch  295/538 - Train Accuracy: 0.9293, Validation Accuracy: 0.9073, Loss: 0.0624
Epoch   2 Batch  300/538 - Train Accuracy: 0.9187, Validation Accuracy: 0.9171, Loss: 0.0657
Epoch   2 Batch  305/538 - Train Accuracy: 0.9234, Validation Accuracy: 0.9299, Loss: 0.0558
Epoch   2 Batch  310/538 - Train Accuracy: 0.9365, Validation Accuracy: 0.9160, Loss: 0.0704
Epoch   2 Batch  315/538 - Train Accuracy: 0.9074, Validation Accuracy: 0.9237, Loss: 0.0648
Epoch   2 Batch  320/538 - Train Accuracy: 0.9185, Validation Accuracy

Epoch   3 Batch  180/538 - Train Accuracy: 0.9291, Validation Accuracy: 0.9057, Loss: 0.0579
Epoch   3 Batch  185/538 - Train Accuracy: 0.9355, Validation Accuracy: 0.9221, Loss: 0.0528
Epoch   3 Batch  190/538 - Train Accuracy: 0.9293, Validation Accuracy: 0.9210, Loss: 0.0727
Epoch   3 Batch  195/538 - Train Accuracy: 0.9349, Validation Accuracy: 0.9212, Loss: 0.0686
Epoch   3 Batch  200/538 - Train Accuracy: 0.9436, Validation Accuracy: 0.9139, Loss: 0.0525
Epoch   3 Batch  205/538 - Train Accuracy: 0.9345, Validation Accuracy: 0.9256, Loss: 0.0547
Epoch   3 Batch  210/538 - Train Accuracy: 0.9211, Validation Accuracy: 0.9435, Loss: 0.0581
Epoch   3 Batch  215/538 - Train Accuracy: 0.9313, Validation Accuracy: 0.9281, Loss: 0.0472
Epoch   3 Batch  220/538 - Train Accuracy: 0.9252, Validation Accuracy: 0.9123, Loss: 0.0579
Epoch   3 Batch  225/538 - Train Accuracy: 0.9343, Validation Accuracy: 0.9134, Loss: 0.0514
Epoch   3 Batch  230/538 - Train Accuracy: 0.9186, Validation Accuracy

Epoch   4 Batch   90/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9201, Loss: 0.0512
Epoch   4 Batch   95/538 - Train Accuracy: 0.9297, Validation Accuracy: 0.9389, Loss: 0.0482
Epoch   4 Batch  100/538 - Train Accuracy: 0.9352, Validation Accuracy: 0.9324, Loss: 0.0459
Epoch   4 Batch  105/538 - Train Accuracy: 0.9345, Validation Accuracy: 0.9382, Loss: 0.0404
Epoch   4 Batch  110/538 - Train Accuracy: 0.9320, Validation Accuracy: 0.9313, Loss: 0.0465
Epoch   4 Batch  115/538 - Train Accuracy: 0.9467, Validation Accuracy: 0.9355, Loss: 0.0489
Epoch   4 Batch  120/538 - Train Accuracy: 0.9498, Validation Accuracy: 0.9300, Loss: 0.0403
Epoch   4 Batch  125/538 - Train Accuracy: 0.9321, Validation Accuracy: 0.9283, Loss: 0.0602
Epoch   4 Batch  130/538 - Train Accuracy: 0.9416, Validation Accuracy: 0.9293, Loss: 0.0494
Epoch   4 Batch  135/538 - Train Accuracy: 0.9276, Validation Accuracy: 0.9359, Loss: 0.0689
Epoch   4 Batch  140/538 - Train Accuracy: 0.9164, Validation Accuracy

Epoch   4 Batch  535/538 - Train Accuracy: 0.9503, Validation Accuracy: 0.9331, Loss: 0.0442
Epoch   5 Batch    5/538 - Train Accuracy: 0.9405, Validation Accuracy: 0.9357, Loss: 0.0574
Epoch   5 Batch   10/538 - Train Accuracy: 0.9406, Validation Accuracy: 0.9389, Loss: 0.0533
Epoch   5 Batch   15/538 - Train Accuracy: 0.9433, Validation Accuracy: 0.9450, Loss: 0.0500
Epoch   5 Batch   20/538 - Train Accuracy: 0.9310, Validation Accuracy: 0.9466, Loss: 0.0511
Epoch   5 Batch   25/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9499, Loss: 0.0517
Epoch   5 Batch   30/538 - Train Accuracy: 0.9383, Validation Accuracy: 0.9501, Loss: 0.0515
Epoch   5 Batch   35/538 - Train Accuracy: 0.9410, Validation Accuracy: 0.9419, Loss: 0.0386
Epoch   5 Batch   40/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9419, Loss: 0.0400
Epoch   5 Batch   45/538 - Train Accuracy: 0.9343, Validation Accuracy: 0.9423, Loss: 0.0534
Epoch   5 Batch   50/538 - Train Accuracy: 0.9461, Validation Accuracy

Epoch   5 Batch  445/538 - Train Accuracy: 0.9480, Validation Accuracy: 0.9379, Loss: 0.0397
Epoch   5 Batch  450/538 - Train Accuracy: 0.9269, Validation Accuracy: 0.9339, Loss: 0.0567
Epoch   5 Batch  455/538 - Train Accuracy: 0.9467, Validation Accuracy: 0.9396, Loss: 0.0520
Epoch   5 Batch  460/538 - Train Accuracy: 0.9267, Validation Accuracy: 0.9444, Loss: 0.0515
Epoch   5 Batch  465/538 - Train Accuracy: 0.9186, Validation Accuracy: 0.9464, Loss: 0.0501
Epoch   5 Batch  470/538 - Train Accuracy: 0.9327, Validation Accuracy: 0.9453, Loss: 0.0501
Epoch   5 Batch  475/538 - Train Accuracy: 0.9408, Validation Accuracy: 0.9411, Loss: 0.0443
Epoch   5 Batch  480/538 - Train Accuracy: 0.9492, Validation Accuracy: 0.9237, Loss: 0.0414
Epoch   5 Batch  485/538 - Train Accuracy: 0.9466, Validation Accuracy: 0.9189, Loss: 0.0575
Epoch   5 Batch  490/538 - Train Accuracy: 0.9448, Validation Accuracy: 0.9336, Loss: 0.0470
Epoch   5 Batch  495/538 - Train Accuracy: 0.9357, Validation Accuracy

Epoch   6 Batch  355/538 - Train Accuracy: 0.9488, Validation Accuracy: 0.9519, Loss: 0.0499
Epoch   6 Batch  360/538 - Train Accuracy: 0.9398, Validation Accuracy: 0.9556, Loss: 0.0489
Epoch   6 Batch  365/538 - Train Accuracy: 0.9422, Validation Accuracy: 0.9471, Loss: 0.0528
Epoch   6 Batch  370/538 - Train Accuracy: 0.9500, Validation Accuracy: 0.9352, Loss: 0.0436
Epoch   6 Batch  375/538 - Train Accuracy: 0.9550, Validation Accuracy: 0.9498, Loss: 0.0480
Epoch   6 Batch  380/538 - Train Accuracy: 0.9434, Validation Accuracy: 0.9471, Loss: 0.0560
Epoch   6 Batch  385/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9485, Loss: 0.0475
Epoch   6 Batch  390/538 - Train Accuracy: 0.9288, Validation Accuracy: 0.9302, Loss: 0.0398
Epoch   6 Batch  395/538 - Train Accuracy: 0.9131, Validation Accuracy: 0.9268, Loss: 0.0524
Epoch   6 Batch  400/538 - Train Accuracy: 0.9276, Validation Accuracy: 0.9435, Loss: 0.0456
Epoch   6 Batch  405/538 - Train Accuracy: 0.9498, Validation Accuracy

Epoch   7 Batch  265/538 - Train Accuracy: 0.9268, Validation Accuracy: 0.9450, Loss: 0.0589
Epoch   7 Batch  270/538 - Train Accuracy: 0.9426, Validation Accuracy: 0.9510, Loss: 0.0416
Epoch   7 Batch  275/538 - Train Accuracy: 0.9348, Validation Accuracy: 0.9391, Loss: 0.0555
Epoch   7 Batch  280/538 - Train Accuracy: 0.9449, Validation Accuracy: 0.9364, Loss: 0.0403
Epoch   7 Batch  285/538 - Train Accuracy: 0.9581, Validation Accuracy: 0.9222, Loss: 0.0389
Epoch   7 Batch  290/538 - Train Accuracy: 0.9574, Validation Accuracy: 0.9395, Loss: 0.0424
Epoch   7 Batch  295/538 - Train Accuracy: 0.9496, Validation Accuracy: 0.9419, Loss: 0.0396
Epoch   7 Batch  300/538 - Train Accuracy: 0.9446, Validation Accuracy: 0.9451, Loss: 0.0496
Epoch   7 Batch  305/538 - Train Accuracy: 0.9487, Validation Accuracy: 0.9554, Loss: 0.0404
Epoch   7 Batch  310/538 - Train Accuracy: 0.9561, Validation Accuracy: 0.9522, Loss: 0.0571
Epoch   7 Batch  315/538 - Train Accuracy: 0.9325, Validation Accuracy

Epoch   8 Batch  175/538 - Train Accuracy: 0.9439, Validation Accuracy: 0.9585, Loss: 0.0395
Epoch   8 Batch  180/538 - Train Accuracy: 0.9541, Validation Accuracy: 0.9416, Loss: 0.0538
Epoch   8 Batch  185/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9469, Loss: 0.0397
Epoch   8 Batch  190/538 - Train Accuracy: 0.9366, Validation Accuracy: 0.9462, Loss: 0.0569
Epoch   8 Batch  195/538 - Train Accuracy: 0.9466, Validation Accuracy: 0.9426, Loss: 0.0558
Epoch   8 Batch  200/538 - Train Accuracy: 0.9646, Validation Accuracy: 0.9430, Loss: 0.0319
Epoch   8 Batch  205/538 - Train Accuracy: 0.9360, Validation Accuracy: 0.9466, Loss: 0.0442
Epoch   8 Batch  210/538 - Train Accuracy: 0.9371, Validation Accuracy: 0.9524, Loss: 0.0515
Epoch   8 Batch  215/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9489, Loss: 0.0458
Epoch   8 Batch  220/538 - Train Accuracy: 0.9349, Validation Accuracy: 0.9446, Loss: 0.0484
Epoch   8 Batch  225/538 - Train Accuracy: 0.9533, Validation Accuracy

Epoch   9 Batch   85/538 - Train Accuracy: 0.9577, Validation Accuracy: 0.9483, Loss: 0.0457
Epoch   9 Batch   90/538 - Train Accuracy: 0.9550, Validation Accuracy: 0.9398, Loss: 0.0427
Epoch   9 Batch   95/538 - Train Accuracy: 0.9318, Validation Accuracy: 0.9512, Loss: 0.0420
Epoch   9 Batch  100/538 - Train Accuracy: 0.9410, Validation Accuracy: 0.9462, Loss: 0.0405
Epoch   9 Batch  105/538 - Train Accuracy: 0.9451, Validation Accuracy: 0.9407, Loss: 0.0369
Epoch   9 Batch  110/538 - Train Accuracy: 0.9504, Validation Accuracy: 0.9467, Loss: 0.0383
Epoch   9 Batch  115/538 - Train Accuracy: 0.9553, Validation Accuracy: 0.9389, Loss: 0.0381
Epoch   9 Batch  120/538 - Train Accuracy: 0.9541, Validation Accuracy: 0.9437, Loss: 0.0351
Epoch   9 Batch  125/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9393, Loss: 0.0428
Epoch   9 Batch  130/538 - Train Accuracy: 0.9433, Validation Accuracy: 0.9338, Loss: 0.0416
Epoch   9 Batch  135/538 - Train Accuracy: 0.9302, Validation Accuracy

Epoch   9 Batch  530/538 - Train Accuracy: 0.9350, Validation Accuracy: 0.9521, Loss: 0.0424
Epoch   9 Batch  535/538 - Train Accuracy: 0.9461, Validation Accuracy: 0.9494, Loss: 0.0402
Epoch  10 Batch    5/538 - Train Accuracy: 0.9449, Validation Accuracy: 0.9387, Loss: 0.0559
Epoch  10 Batch   10/538 - Train Accuracy: 0.9379, Validation Accuracy: 0.9396, Loss: 0.0570
Epoch  10 Batch   15/538 - Train Accuracy: 0.9528, Validation Accuracy: 0.9535, Loss: 0.0425
Epoch  10 Batch   20/538 - Train Accuracy: 0.9317, Validation Accuracy: 0.9453, Loss: 0.0502
Epoch  10 Batch   25/538 - Train Accuracy: 0.9381, Validation Accuracy: 0.9533, Loss: 0.0562
Epoch  10 Batch   30/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9577, Loss: 0.0506
Epoch  10 Batch   35/538 - Train Accuracy: 0.9426, Validation Accuracy: 0.9595, Loss: 0.0433
Epoch  10 Batch   40/538 - Train Accuracy: 0.9524, Validation Accuracy: 0.9583, Loss: 0.0363
Epoch  10 Batch   45/538 - Train Accuracy: 0.9362, Validation Accuracy

Epoch  10 Batch  440/538 - Train Accuracy: 0.9309, Validation Accuracy: 0.9263, Loss: 0.0589
Epoch  10 Batch  445/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9339, Loss: 0.0381
Epoch  10 Batch  450/538 - Train Accuracy: 0.9295, Validation Accuracy: 0.9409, Loss: 0.0568
Epoch  10 Batch  455/538 - Train Accuracy: 0.9466, Validation Accuracy: 0.9395, Loss: 0.0495
Epoch  10 Batch  460/538 - Train Accuracy: 0.9317, Validation Accuracy: 0.9535, Loss: 0.0566
Epoch  10 Batch  465/538 - Train Accuracy: 0.9492, Validation Accuracy: 0.9375, Loss: 0.0474
Epoch  10 Batch  470/538 - Train Accuracy: 0.9513, Validation Accuracy: 0.9487, Loss: 0.0405
Epoch  10 Batch  475/538 - Train Accuracy: 0.9403, Validation Accuracy: 0.9533, Loss: 0.0508
Epoch  10 Batch  480/538 - Train Accuracy: 0.9548, Validation Accuracy: 0.9380, Loss: 0.0450
Epoch  10 Batch  485/538 - Train Accuracy: 0.9537, Validation Accuracy: 0.9435, Loss: 0.0581
Epoch  10 Batch  490/538 - Train Accuracy: 0.9548, Validation Accuracy

Epoch  11 Batch  350/538 - Train Accuracy: 0.9479, Validation Accuracy: 0.9492, Loss: 0.0529
Epoch  11 Batch  355/538 - Train Accuracy: 0.9623, Validation Accuracy: 0.9409, Loss: 0.0409
Epoch  11 Batch  360/538 - Train Accuracy: 0.9375, Validation Accuracy: 0.9544, Loss: 0.0479
Epoch  11 Batch  365/538 - Train Accuracy: 0.9485, Validation Accuracy: 0.9471, Loss: 0.0485
Epoch  11 Batch  370/538 - Train Accuracy: 0.9584, Validation Accuracy: 0.9439, Loss: 0.0370
Epoch  11 Batch  375/538 - Train Accuracy: 0.9641, Validation Accuracy: 0.9389, Loss: 0.0486
Epoch  11 Batch  380/538 - Train Accuracy: 0.9594, Validation Accuracy: 0.9474, Loss: 0.0432
Epoch  11 Batch  385/538 - Train Accuracy: 0.9598, Validation Accuracy: 0.9506, Loss: 0.0473
Epoch  11 Batch  390/538 - Train Accuracy: 0.9433, Validation Accuracy: 0.9531, Loss: 0.0365
Epoch  11 Batch  395/538 - Train Accuracy: 0.9475, Validation Accuracy: 0.9526, Loss: 0.0524
Epoch  11 Batch  400/538 - Train Accuracy: 0.9585, Validation Accuracy

Epoch  12 Batch  260/538 - Train Accuracy: 0.9195, Validation Accuracy: 0.9551, Loss: 0.0513
Epoch  12 Batch  265/538 - Train Accuracy: 0.9375, Validation Accuracy: 0.9466, Loss: 0.0553
Epoch  12 Batch  270/538 - Train Accuracy: 0.9514, Validation Accuracy: 0.9631, Loss: 0.0395
Epoch  12 Batch  275/538 - Train Accuracy: 0.9354, Validation Accuracy: 0.9496, Loss: 0.0522
Epoch  12 Batch  280/538 - Train Accuracy: 0.9397, Validation Accuracy: 0.9402, Loss: 0.0325
Epoch  12 Batch  285/538 - Train Accuracy: 0.9578, Validation Accuracy: 0.9352, Loss: 0.0431
Epoch  12 Batch  290/538 - Train Accuracy: 0.9564, Validation Accuracy: 0.9313, Loss: 0.0406
Epoch  12 Batch  295/538 - Train Accuracy: 0.9503, Validation Accuracy: 0.9403, Loss: 0.0491
Epoch  12 Batch  300/538 - Train Accuracy: 0.9487, Validation Accuracy: 0.9570, Loss: 0.0502
Epoch  12 Batch  305/538 - Train Accuracy: 0.9505, Validation Accuracy: 0.9609, Loss: 0.0412
Epoch  12 Batch  310/538 - Train Accuracy: 0.9596, Validation Accuracy

Epoch  13 Batch  170/538 - Train Accuracy: 0.9444, Validation Accuracy: 0.9455, Loss: 0.0458
Epoch  13 Batch  175/538 - Train Accuracy: 0.9533, Validation Accuracy: 0.9416, Loss: 0.0491
Epoch  13 Batch  180/538 - Train Accuracy: 0.9611, Validation Accuracy: 0.9325, Loss: 0.0452
Epoch  13 Batch  185/538 - Train Accuracy: 0.9719, Validation Accuracy: 0.9483, Loss: 0.0360
Epoch  13 Batch  190/538 - Train Accuracy: 0.9446, Validation Accuracy: 0.9462, Loss: 0.0549
Epoch  13 Batch  195/538 - Train Accuracy: 0.9580, Validation Accuracy: 0.9308, Loss: 0.0511
Epoch  13 Batch  200/538 - Train Accuracy: 0.9633, Validation Accuracy: 0.9348, Loss: 0.0341
Epoch  13 Batch  205/538 - Train Accuracy: 0.9431, Validation Accuracy: 0.9469, Loss: 0.0448
Epoch  13 Batch  210/538 - Train Accuracy: 0.9355, Validation Accuracy: 0.9405, Loss: 0.0566
Epoch  13 Batch  215/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9370, Loss: 0.0471
Epoch  13 Batch  220/538 - Train Accuracy: 0.9435, Validation Accuracy

Epoch  14 Batch   80/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9544, Loss: 0.0396
Epoch  14 Batch   85/538 - Train Accuracy: 0.9613, Validation Accuracy: 0.9595, Loss: 0.0408
Epoch  14 Batch   90/538 - Train Accuracy: 0.9591, Validation Accuracy: 0.9576, Loss: 0.0443
Epoch  14 Batch   95/538 - Train Accuracy: 0.9288, Validation Accuracy: 0.9522, Loss: 0.0511
Epoch  14 Batch  100/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9522, Loss: 0.0448
Epoch  14 Batch  105/538 - Train Accuracy: 0.9446, Validation Accuracy: 0.9533, Loss: 0.0392
Epoch  14 Batch  110/538 - Train Accuracy: 0.9580, Validation Accuracy: 0.9583, Loss: 0.0461
Epoch  14 Batch  115/538 - Train Accuracy: 0.9543, Validation Accuracy: 0.9448, Loss: 0.0433
Epoch  14 Batch  120/538 - Train Accuracy: 0.9564, Validation Accuracy: 0.9503, Loss: 0.0366
Epoch  14 Batch  125/538 - Train Accuracy: 0.9479, Validation Accuracy: 0.9545, Loss: 0.0593
Epoch  14 Batch  130/538 - Train Accuracy: 0.9488, Validation Accuracy

Epoch  14 Batch  525/538 - Train Accuracy: 0.9528, Validation Accuracy: 0.9450, Loss: 0.0486
Epoch  14 Batch  530/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9467, Loss: 0.0559
Epoch  14 Batch  535/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9508, Loss: 0.0454
Epoch  15 Batch    5/538 - Train Accuracy: 0.9485, Validation Accuracy: 0.9474, Loss: 0.0466
Epoch  15 Batch   10/538 - Train Accuracy: 0.9525, Validation Accuracy: 0.9428, Loss: 0.0510
Epoch  15 Batch   15/538 - Train Accuracy: 0.9448, Validation Accuracy: 0.9421, Loss: 0.0501
Epoch  15 Batch   20/538 - Train Accuracy: 0.9565, Validation Accuracy: 0.9512, Loss: 0.0477
Epoch  15 Batch   25/538 - Train Accuracy: 0.9244, Validation Accuracy: 0.9407, Loss: 0.0566
Epoch  15 Batch   30/538 - Train Accuracy: 0.9303, Validation Accuracy: 0.9455, Loss: 0.0563
Epoch  15 Batch   35/538 - Train Accuracy: 0.9410, Validation Accuracy: 0.9442, Loss: 0.0390
Epoch  15 Batch   40/538 - Train Accuracy: 0.9531, Validation Accuracy

Epoch  15 Batch  435/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9482, Loss: 0.0514
Epoch  15 Batch  440/538 - Train Accuracy: 0.9322, Validation Accuracy: 0.9482, Loss: 0.0614
Epoch  15 Batch  445/538 - Train Accuracy: 0.9535, Validation Accuracy: 0.9549, Loss: 0.0483
Epoch  15 Batch  450/538 - Train Accuracy: 0.9353, Validation Accuracy: 0.9565, Loss: 0.0573
Epoch  15 Batch  455/538 - Train Accuracy: 0.9528, Validation Accuracy: 0.9545, Loss: 0.0446
Epoch  15 Batch  460/538 - Train Accuracy: 0.9388, Validation Accuracy: 0.9467, Loss: 0.0477
Epoch  15 Batch  465/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9439, Loss: 0.0481
Epoch  15 Batch  470/538 - Train Accuracy: 0.9418, Validation Accuracy: 0.9384, Loss: 0.0432
Epoch  15 Batch  475/538 - Train Accuracy: 0.9477, Validation Accuracy: 0.9462, Loss: 0.0506
Epoch  15 Batch  480/538 - Train Accuracy: 0.9427, Validation Accuracy: 0.9498, Loss: 0.0475
Epoch  15 Batch  485/538 - Train Accuracy: 0.9492, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [24]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    unknown = vocab_to_int['<UNK>']
    return [vocab_to_int.get(word, unknown) for word in sentence.lower().split()]



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [25]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [156, 52, 83, 167, 134, 80, 144]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [30, 125, 115, 25, 87, 88, 57, 140, 1]
  French Words: il a vu un vieux camion rouge . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.